File: .github/scripts/generate_yaml.go
package main

import (
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"net/http"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"sync"
	"time"
)

const (
	distDir   = "dist/yaml"
	configDir = "config"
	cacheDir  = ".cache/external"
)

// === Caching Logic ===

func getCachePath(url string) string {
	hash := sha256.Sum256([]byte(url))
	return filepath.Join(cacheDir, hex.EncodeToString(hash[:]))
}

type CacheMeta struct {
	ETag         string `json:"etag"`
	LastModified string `json:"last_modified"`
}

func fetchURL(url string) ([]string, error) {
	if url == "" {
		return nil, nil
	}

	cacheFile := getCachePath(url)
	metaFile := cacheFile + ".meta"

	// Ensure cache directory exists
	if err := os.MkdirAll(cacheDir, 0755); err != nil {
		return nil, err
	}

	// Prepare request
	req, err := http.NewRequest("GET", url, nil)
	if err != nil {
		return nil, err
	}
	req.Header.Set("User-Agent", "iplist-bot/1.0")

	// Load cache metadata
	var meta CacheMeta
	if data, err := os.ReadFile(metaFile); err == nil {
		json.Unmarshal(data, &meta)
		if meta.ETag != "" {
			req.Header.Set("If-None-Match", meta.ETag)
		}
		if meta.LastModified != "" {
			req.Header.Set("If-Modified-Since", meta.LastModified)
		}
	}

	// Do request
	client := &http.Client{Timeout: 10 * time.Second}
	resp, err := client.Do(req)
	if err != nil {
		// On error, if we have cache, fallback to it
		if _, err := os.Stat(cacheFile); err == nil {
			log.Printf("Network error fetching %s, using outdated cache: %v", url, err)
			return readLines(cacheFile)
		}
		return nil, err
	}
	defer resp.Body.Close()

	// Handle Response
	if resp.StatusCode == http.StatusNotModified {
		log.Printf("Cached (304): %s", url)
		return readLines(cacheFile)
	}

	if resp.StatusCode != http.StatusOK {
		// Logic: if fail, fallback to cache if valid
		if _, err := os.Stat(cacheFile); err == nil {
			log.Printf("HTTP %d fetching %s, using outdated cache", resp.StatusCode, url)
			return readLines(cacheFile)
		}
		return nil, fmt.Errorf("HTTP %d", resp.StatusCode)
	}

	// Read body
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}

	// Save to cache
	if err := os.WriteFile(cacheFile, body, 0644); err != nil {
		log.Printf("Failed to write cache: %v", err)
	}

	// Save metadata
	newMeta := CacheMeta{
		ETag:         resp.Header.Get("ETag"),
		LastModified: resp.Header.Get("Last-Modified"),
	}
	if metaBytes, err := json.Marshal(newMeta); err == nil {
		os.WriteFile(metaFile, metaBytes, 0644)
	}

	log.Printf("Fetched (200): %s", url)
	return strings.Split(string(body), "\n"), nil
}

func readLines(path string) ([]string, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}
	return strings.Split(string(data), "\n"), nil
}

// === Normalization ===

func normalizeDomain(line string) string {
	line = strings.TrimSpace(line)
	line = strings.ReplaceAll(line, "\"", "")
	if line == "" || strings.HasPrefix(line, "#") {
		return ""
	}
	// If no wildcard (*), no +., start with dot -> assume domain -> add +.
	// Logic from python script:
	// if not (line.startswith('*') or line.startswith('+.') or line.startswith('.')) and '/' not in line
	if !strings.HasPrefix(line, "*") && !strings.HasPrefix(line, "+.") && !strings.HasPrefix(line, ".") && !strings.Contains(line, "/") {
		return "+." + line
	}
	return line
}

func normalizeIP(line string) string {
	line = strings.TrimSpace(line)
	line = strings.ReplaceAll(line, "\"", "")
	if line == "" || strings.HasPrefix(line, "#") {
		return ""
	}
	// If already has CIDR
	if strings.Contains(line, "/") {
		return line
	}
	// Add CIDR
	if strings.Contains(line, ":") {
		return line + "/128"
	}
	return line + "/32"
}

// === Config Processing ===

type Config struct {
	Domains []string          `json:"domains"`
	IP4     []string          `json:"ip4"`
	IP6     []string          `json:"ip6"`
	CIDR4   []string          `json:"cidr4"`
	CIDR6   []string          `json:"cidr6"`
	External map[string][]string `json:"external"`
}

type Result struct {
	Group   string
	Site    string
	Domains []string
	IPs     []string
}

func processConfig(path string) (*Result, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}

	var cfg Config
	if err := json.Unmarshal(data, &cfg); err != nil {
		return nil, err
	}

	relPath, _ := filepath.Rel(configDir, path)
	group := filepath.Dir(relPath)
	site := strings.TrimSuffix(filepath.Base(relPath), filepath.Ext(path))

	if group == "." || site == "" {
		return nil, nil // skip root files
	}

	var domains []string
	var ips []string

	// Local data
	domains = append(domains, cfg.Domains...)
	ips = append(ips, cfg.IP4...)
	ips = append(ips, cfg.IP6...)
	ips = append(ips, cfg.CIDR4...)
	ips = append(ips, cfg.CIDR6...)

	// External
	for _, url := range cfg.External["domains"] {
		lines, _ := fetchURL(url)
		domains = append(domains, lines...)
	}
	for _, field := range []string{"ip4", "ip6", "cidr4", "cidr6"} {
		for _, url := range cfg.External[field] {
			lines, _ := fetchURL(url)
			ips = append(ips, lines...)
		}
	}

	// Normalize
	domSet := make(map[string]bool)
	for _, d := range domains {
		n := normalizeDomain(d)
		if n != "" {
			domSet[n] = true
		}
	}

	ipSet := make(map[string]bool)
	for _, i := range ips {
		n := normalizeIP(i)
		if n != "" {
			ipSet[n] = true
		}
	}

	res := &Result{
		Group:   group,
		Site:    site,
		Domains: make([]string, 0, len(domSet)),
		IPs:     make([]string, 0, len(ipSet)),
	}

	for k := range domSet {
		res.Domains = append(res.Domains, k)
	}
	for k := range ipSet {
		res.IPs = append(res.IPs, k)
	}
	sort.Strings(res.Domains)
	sort.Strings(res.IPs)

	return res, nil
}

// === Writing ===

func writeYAML(path string, items []string) {
	if len(items) == 0 {
		return
	}
	// Sort again just in case
	sort.Strings(items)

	// Build content
	var sb strings.Builder
	sb.WriteString("payload:\n")
	for _, item := range items {
		sb.WriteString("  - \"")
		sb.WriteString(item)
		sb.WriteString("\"\n")
	}
	newContent := sb.String()

	// Compare with existing
	if existing, err := os.ReadFile(path); err == nil {
		if string(existing) == newContent {
			return // skip
		}
	}

	// Write
	os.WriteFile(path, []byte(newContent), 0644)
}

func main() {
	os.MkdirAll(filepath.Join(distDir, "domain"), 0755)
	os.MkdirAll(filepath.Join(distDir, "ipcidr"), 0755)

	files, err := filepath.Glob(filepath.Join(configDir, "*", "*.json"))
	if err != nil {
		log.Fatal(err)
	}

	results := make([]*Result, len(files))
	var wg sync.WaitGroup

	// Semaphore/worker pool logic (8 workers)
	sem := make(chan struct{}, 8)

	for i, f := range files {
		wg.Add(1)
		go func(i int, f string) {
			defer wg.Done()
			sem <- struct{}{} // acquire
			res, err := processConfig(f)
			<-sem // release
			
			if err != nil {
				log.Printf("Error processing %s: %v", f, err)
			}
			results[i] = res
		}(i, f)
	}
	wg.Wait()

	// Aggregation
	groupsDomains := make(map[string][]string)
	groupsIPs := make(map[string][]string)
	var allDomains []string
	var allIPs []string

	for _, res := range results {
		if res == nil {
			continue
		}

		// Write site files
		writeYAML(filepath.Join(distDir, "domain", fmt.Sprintf("%s--%s.yaml", res.Group, res.Site)), res.Domains)
		writeYAML(filepath.Join(distDir, "ipcidr", fmt.Sprintf("%s--%s.yaml", res.Group, res.Site)), res.IPs)

		// Aggregate
		groupsDomains[res.Group] = append(groupsDomains[res.Group], res.Domains...)
		groupsIPs[res.Group] = append(groupsIPs[res.Group], res.IPs...)
		allDomains = append(allDomains, res.Domains...)
		allIPs = append(allIPs, res.IPs...)
	}

	// Write Groups (deduplicated)
	for g, items := range groupsDomains {
		writeYAML(filepath.Join(distDir, "domain", g+".yaml"), unique(items))
	}
	for g, items := range groupsIPs {
		writeYAML(filepath.Join(distDir, "ipcidr", g+".yaml"), unique(items))
	}

	// Write All (deduplicated)
	writeYAML(filepath.Join(distDir, "domain", "all.yaml"), unique(allDomains))
	writeYAML(filepath.Join(distDir, "ipcidr", "all.yaml"), unique(allIPs))

	fmt.Println("Go YAML generation complete.")
}

func unique(items []string) []string {
	seen := make(map[string]bool)
	var out []string
	for _, s := range items {
		if !seen[s] {
			seen[s] = true
			out = append(out, s)
		}
	}
	return out
}

File: .github/scripts/generate_yaml.py
import os
import glob
import json
import urllib.request
import sys
from concurrent.futures import ThreadPoolExecutor

DIST_DIR = "dist/yaml"
CONFIG_DIR = "config"

def fetch_url(url):
    try:
        if not url: return []
        print(f"Fetching external: {url}")
        # Default timeout 10s
        with urllib.request.urlopen(url, timeout=10) as response:
            return response.read().decode('utf-8').splitlines()
    except Exception as e:
        print(f"Error fetching {url}: {e}", file=sys.stderr)
        return []

def normalize_domain(line):
    line = line.strip()
    # Remove quotes if present
    line = line.replace('"', '')
    if not line or line.startswith('#'):
        return None
    # Logic: if no wildcard/special prefix and no path separator, assume domain -> add +.
    if not (line.startswith('*') or line.startswith('+.') or line.startswith('.')) and '/' not in line:
        return "+." + line
    return line

def _suffix_host(pattern):
    if pattern.startswith('+.'):
        return pattern[2:]
    if pattern.startswith('.'):
        return pattern[1:]
    return None

def prune_subdomains(domains):
    # Drop redundant subdomains when a broader domain-suffix entry exists (e.g. +.example.com -> +.www.example.com).
    suffix_hosts = set()
    for item in domains:
        host = _suffix_host(item)
        if host:
            suffix_hosts.add(host)

    if not suffix_hosts:
        return domains

    redundant_hosts = set()
    for host in suffix_hosts:
        parts = host.split('.')
        for i in range(1, len(parts)):
            parent = '.'.join(parts[i:])
            if parent in suffix_hosts:
                redundant_hosts.add(host)
                break

    if not redundant_hosts:
        return domains

    return {item for item in domains if _suffix_host(item) not in redundant_hosts}

def normalize_ip(line):
    line = line.strip()
    line = line.replace('"', '')
    if not line or line.startswith('#'):
        return None
    
    # If already has CIDR mask, return as-is
    if '/' in line:
        return line
    
    # Add CIDR mask for bare IPs
    if ':' in line:
        # IPv6 - add /128
        return f"{line}/128"
    else:
        # IPv4 - add /32
        return f"{line}/32"

def process_config(config_path):
    try:
        with open(config_path, 'r') as f:
            data = json.load(f)
    except Exception as e:
        print(f"Error reading {config_path}: {e}", file=sys.stderr)
        return None

    rel_path = os.path.relpath(config_path, CONFIG_DIR)
    group = os.path.dirname(rel_path)
    site = os.path.splitext(os.path.basename(rel_path))[0]
    
    if not group or not site:
        return None

    domains = data.get('domains', [])
    ip_list = []
    
    # Collect IPs/CIDRs
    ip_list.extend(data.get('ip4', []))
    ip_list.extend(data.get('ip6', []))
    ip_list.extend(data.get('cidr4', []))
    ip_list.extend(data.get('cidr6', []))
    
    # External processing
    ext = data.get('external', {})
    if ext:
        # We fetch simply here. For huge lists, could be async, but Python threads are fine for I/O.
        # Check specific external fields
        for field in ['domains']:
            for url in ext.get(field, []):
                domains.extend(fetch_url(url))
        for field in ['ip4', 'ip6', 'cidr4', 'cidr6']:
            for url in ext.get(field, []):
                ip_list.extend(fetch_url(url))

    # Normalize
    norm_domains = set()
    for d in domains:
        n = normalize_domain(str(d))
        if n: norm_domains.add(n)
    norm_domains = prune_subdomains(norm_domains)
        
    norm_ips = set()
    for i in ip_list:
        n = normalize_ip(str(i))
        if n: norm_ips.add(n)

    return {
        'group': group,
        'site': site,
        'domains': norm_domains,
        'ips': norm_ips
    }

def write_yaml(path, items):
    if not items: return
    sorted_items = sorted(items)
    
    # Build new content
    lines = ["payload:\n"]
    for item in sorted_items:
        lines.append(f'  - "{item}"\n')
    new_content = "".join(lines)
    
    # Only write if content differs (preserves timestamp for incremental MRS build)
    if os.path.exists(path):
        with open(path, "r") as f:
            existing = f.read()
        if existing == new_content:
            return  # No change, keep original timestamp
    
    with open(path, "w") as f:
        f.write(new_content)

def main():
    os.makedirs(f"{DIST_DIR}/domain", exist_ok=True)
    os.makedirs(f"{DIST_DIR}/ipcidr", exist_ok=True)
    
    files = glob.glob(f"{CONFIG_DIR}/**/*.json", recursive=True)
    
    # Parallel processing of configs (reads + fetching)
    results = []
    with ThreadPoolExecutor(max_workers=8) as executor:
        results = list(executor.map(process_config, files))
    
    # Aggregation
    groups_domains = {}
    groups_ips = {}
    all_domains = set()
    all_ips = set()
    
    for res in results:
        if not res: continue
        
        g = res['group']
        s = res['site']
        doms = res['domains']
        ips = res['ips']
        
        # Write Site YAML
        write_yaml(f"{DIST_DIR}/domain/{g}--{s}.yaml", doms)
        write_yaml(f"{DIST_DIR}/ipcidr/{g}--{s}.yaml", ips)
        
        # Aggregate Group
        if g not in groups_domains: groups_domains[g] = set()
        if g not in groups_ips: groups_ips[g] = set()
        
        groups_domains[g].update(doms)
        groups_ips[g].update(ips)
        
        # Aggregate All
        all_domains.update(doms)
        all_ips.update(ips)
        
    # Write Group YAMLs
    for g in groups_domains:
        write_yaml(f"{DIST_DIR}/domain/{g}.yaml", groups_domains[g])
    for g in groups_ips:
        write_yaml(f"{DIST_DIR}/ipcidr/{g}.yaml", groups_ips[g])
        
    # Write All YAMLs
    write_yaml(f"{DIST_DIR}/domain/all.yaml", all_domains)
    write_yaml(f"{DIST_DIR}/ipcidr/all.yaml", all_ips)
    
    print("YAML generation complete.")

if __name__ == "__main__":
    main()
